allow native and ssh-keygen public key check

Gibheer opened the pull request:
Hi,

this is the first part of fixing #2179, namely the SELinux and ssh-keygen problem. The first commit is pretty large, but adding public keys now works again using ssh-keygen.

Please do not merge yet, as there is still some stuff missing. I introduced new settings to make it possible to adjust the working directory for ssh-keygen and even adjust the ssh-keygen path. I have to update all occurrences of ssh-keygen for that still.
But if possible, I welcome everyone to test the patch :)

Unknwon commented:
PS: build is failing...

Gibheer commented:
Now this error I have no idea how to fix that. I tried with openssh 5.9 and that returns "is not a public key file" for ed25519 public keys. But what is travis using in this case, that it returns "file too long"?

I would like to keep the tests, but if it breaks the build, it is kind of stupid. Should I delete them or should I just remove the ed25519 key from the test?

Unknwon commented:
I guess we need special instructions to tell Travis (but maybe not possible?) support ed25519 key.

Is this PR finished now?

Gibheer commented:
Not yet. I want to adjust all occurences of ssh-keygen to use the new variables. But if you want, you could pull this request in and the other stuff will be done extra.

Unknwon commented on models/ssh_key.go:
```diff
@@ -33,7 +35,10 @@ const (
-var sshOpLocker = sync.Mutex{}
+var (
+	sshOpLocker          = sync.Mutex{}
+	SSH_UNKNOWN_KEY_TYPE = fmt.Errorf("unknown key type")
```
Variable should not use ALL_CAPS.

Unknwon commented:
What is the different between this PR and revert old approach back?

Gibheer commented:
This is a combination of the old behavior and the new one. At the end it should be possible to use the go native ssh library only without any dependency on ssh-keygen at all. Currently, there are still some places in need of ssh-keygen.
That would also mean, that depending on what server you choose to use (sshd, gogs), you would get the full support from the specific system.

Unknwon commented on modules/setting/setting.go:
```diff
@@ -30,9 +30,11 @@ import (
+	HTTP                       Scheme = "http"
+	HTTPS                      Scheme = "https"
+	FCGI                       Scheme = "fcgi"
+	SSH_PUBLICKEY_CHECK_NATIVE        = "native"
```
This `const` section is for `Scheme`, do not mixed up with them.

Unknwon commented on modules/setting/setting.go:
```diff
@@ -328,6 +333,29 @@ func NewContext() {
+	checkDefault := SSH_PUBLICKEY_CHECK_KEYGEN
+	if DisableSSH {
+		checkDefault = SSH_PUBLICKEY_CHECK_NATIVE
+	}
```
When disable SSH, why should we care the method at all?

Unknwon commented:
ping @Gibheer

Gibheer commented:
@Unknwon sorry, took a while longer. I fixed the stuff you found and replaced the ssh-keygen call with the new variable from setting. You can merge this for now, as I do not know when I have the time to do the rest (making ssh-keygen only necessary when the system sshd is used).
Thank you for your help :)

Unknwon commented:
Thanks!
